src/Simplex/Messaging/Agent.hs

```diff
@@ -1979,6 +2134,48 @@ processSMPTransmission c@AgentClient {smpClients, subQ} (tSess@(_, srv, _), v, s
           | internalPrevMsgHash /= receivedPrevMsgHash = MsgError MsgBadHash
           | otherwise = MsgError MsgDuplicate -- this case is not possible
 
+withSwitchedRQ :: AgentMonad m => AgentClient -> RcvQueue -> RcvSwitchStatus -> (DB.Connection -> RcvQueue -> IO a) -> m (Either QueueCheckError a)
+withSwitchedRQ c rq expected =
+  withExistingRQ c rq (\RcvQueue {switchStatus} -> switchStatus == Just expected)
+
+withExistingRQ :: AgentMonad m => AgentClient -> RcvQueue -> (RcvQueue -> Bool) -> (DB.Connection -> RcvQueue -> IO a) -> m (Either QueueCheckError a)
+withExistingRQ c RcvQueue {connId, dbQueueId} qp action =
+  withStore c $ \db ->
+    getRcvQueueById db connId dbQueueId >>= \case
+      Right rq ->
+        if qp rq
+          then do
+            r <- action db rq
+            pure $ Right (Right r)
+          else pure $ Right (Left QCEPredicateFailed)
+      _ -> pure $ Right (Left QCEQueueNotFound)
+
+withSwitchedSQ :: AgentMonad m => AgentClient -> SndQueue -> SndSwitchStatus -> (DB.Connection -> SndQueue -> IO a) -> m (Either QueueCheckError a)
+withSwitchedSQ c sq expected =
+  withExistingSQ c sq (\SndQueue {switchStatus} -> switchStatus == Just expected)
+
+withExistingSQ :: AgentMonad m => AgentClient -> SndQueue -> (SndQueue -> Bool) -> (DB.Connection -> SndQueue -> IO a) -> m (Either QueueCheckError a)
+withExistingSQ c SndQueue {connId, dbQueueId} qp action =
```

Comment:
same here, just remove this function - there is no need to read the queue again.